File: 2020/day06_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: str) -> Any:
    return sum(len(every_yes(group.split("\n"))) for group in content.split("\n\n"))


def every_yes(group: list[str]):
    return [
        answer
        for answer in set("".join(group))
        if all(answer in person for person in group)
    ]


def main() -> None:
    with open("day06.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()

File: 2020/day06-2.ts
function solution(content: string) {
  return content
    .split("\n\n")
    .map((group) => everyYes(group.split("\n")).length)
    .reduce((acc, cur) => acc + cur, 0);
}

function everyYes(group: string[]) {
  return Array.from(new Set(group.join(""))).filter((answer) =>
    group.every((person) => person.includes(answer))
  );
}

Deno.readTextFile("day06.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
